cql3/expr/expression.cc

```diff
@@ -1265,6 +1222,67 @@ expression replace_token(const expression& expr, const column_definition* new_cd
         }, expr);
 }
 
+expression search_and_replace(const expression& e,
+        const noncopyable_function<std::optional<expression> (const expression& candidate)>& replace_candidate) {
+    //auto recurse = std::bind(search_and_replace, std::placeholders::_1, std::ref(replace_candidate));
+    auto recurse = [&] (const expression& e) -> expression {
+        return search_and_replace(e, replace_candidate);
+    };
+    return replace_candidate(e).value_or(
```

Comment:
`value_or` takes the alternative value by value. That would mean that even when `replace_candidate` returns some value, we are still going to evaluate the alternative option. This doesn't seem good for performance.

Reply:
Right